internal/argocd/revision.go

```diff
@@ -10,68 +11,163 @@ import (
+		if s == nil {
+			err := errors.New("Single-source application source is nil")
+			appLogger.Error(err, "Cannot determine desired revision for application, bailing.")
+			return nil, nil
+		}
+
+		appLogger.Debug("Application source is not nil, checking.", "source", s)
 
 		// If there is a source update that targets app.Spec.Source, it might
 		// have its own ideas about the desired revision.
-		var targetPromoMechanism any
-		for i := range update.SourceUpdates {
-			sourceUpdate := &update.SourceUpdates[i]
-			if sourceUpdate.RepoURL == app.Spec.Source.RepoURL && sourceUpdate.Chart == app.Spec.Source.Chart {
-				targetPromoMechanism = sourceUpdate
-				break
-			}
+		targetUpdate := getTargetUpdate(ctx, update, s)
+		desiredOrigin := freight.GetDesiredOrigin(ctx, stage, targetUpdate)
+		if desiredOrigin == nil {
+			appLogger.WithValues("source", s,
+				"revision", app.Status.Sync.Revision).Debug("Could not determine desired origin " +
+				"for application source, using existing revision.")
+			return []string{app.Status.Sync.Revision}, nil
 		}
-		if targetPromoMechanism == nil {
-			targetPromoMechanism = update
+
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, s, desiredOrigin, frght)
+
+		if err != nil {
+			return nil, err
 		}
-		desiredOrigin := freight.GetDesiredOrigin(stage, targetPromoMechanism)
-		repoURL := app.Spec.Source.RepoURL
-		chartName := app.Spec.Source.Chart
-		if !strings.Contains(repoURL, "://") {
-			// In Argo CD ApplicationSource, if a repo URL specifies no protocol and a
-			// chart name is set (already confirmed at this point), we can assume that
-			// the repo URL is an OCI registry URL. Kargo Warehouses and Freight,
-			// however, do use oci:// at the beginning of such URLs.
-			//
-			// Additionally, where OCI is concerned, an ApplicationSource's repoURL is
-			// really a registry URL, and the chart name is a repository within that
-			// registry. Warehouses and Freight, however, handle things more correctly
-			// where a repoURL points directly to a repository and chart name is
-			// irrelevant / blank. We need to account for this when we search our
-			// Freight for the chart.
-			repoURL = fmt.Sprintf(
-				"oci://%s/%s",
-				strings.TrimSuffix(repoURL, "/"),
-				chartName,
-			)
-			chartName = ""
+
+		if desiredRevision != "" {
+			appLogger.WithValues("source", s,
+				"revision", desiredRevision).Debug("Found desired revision for application source.")
+			return []string{desiredRevision}, nil
 		}
+
+		appLogger.WithValues("source", s).Debug("Could not determine desired revision for application from this source.")
+		return nil, nil
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
```

Comment:
Is this comment relevant? This seems like it might be a remnant an earlier iteration of this PR.